kernel/posix/pthread_semaphore.c

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pthread.h>
+
+typedef struct pthread_sem {
+	struct k_sem *sem;
+} sem_t;
+
+int sem_destroy(sem_t *semaphore)
+{
+	k_sem_reset(semaphore);
+	return 0;
+}
+
+int sem_getvalue(sem_t *semaphore, int *value)
+{
+	int count;
+
+	count = k_sem_count_get(semaphore);
+	*value = count;
+
+	return 0;
+}
+
+int sem_init(sem_t *semaphore, int pshared, unsigned value)
+{
+	if (semaphore->limit = 0){
+		/* limit can not be zero*/
+		return EINVAL;
+	}
+	ARG_UNUSED(pshared);
+	k_sem_init(semaphore, value, semaphore->limit);
+
+	return 0;
+}
+
+int sem_post(sem_t *semaphore)
+{
+	k_sem_give(semaphore);
+	return 0;
+}
+
+int sem_timedwait(sem_t *semaphore, const struct timespec *abstime)
+{
+	u32_t 		timeout;
+	long		msecs = 0;
+	int 		retc = 0;
+	s64_t		current_ms;
+	int		abstime_ms;
+	bool		retval;
+
+	if ((abstime->tv_sec < 0) || (1000000000 <= abstime->tv_nsec)) {
+		return -1;
+	}
+
+	current_ms = k_uptime_get();
+	abstime_ms = _ts_to_ms(abstime);
```

Comment:
Can any `abstime` value overflow an `int`?

Reply:
No it wont overflow here because for real time system, maximum value of seconds and nano seconds are enough within s32_t range.  abstime represents how time user want to wait on semaphore for acquiring it in terms of seconds and nano seconds.

Reply:
I wouldn't be worried if abstime were a relative time, but it's an absolute time (seconds+nanoseconds since the Unix epoch), and a 32-bit integer will overflow in 2038.

Reply:
This still needs to be addressed.